pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,623 @@
+		}
+
+		return nil, fmt.Errorf("failed to delete old rule: %w", err)
+	}
+
+	return routes, nil
+}
+
+const (
+	upgradeRevertWarning = "Reverting the new ENI datapath failed. However, the previous datapath is still intact. " +
+		"Endpoint connectivity should not be affected. It is advised to retry the migration."
+	downgradeRevertWarning = "Reverting the new ENI datapath failed. However, both the new and previous datapaths are still intact. " +
+		"Endpoint connectivity should not be affected. It is advised to retry the migration."
+)
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. The table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex, plus the specific table offset
+// value.
+func (m *migrator) retrieveTableIDFromIfIndex(ifindex int) (int, error) {
+	link, err := m.rpdb.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := m.getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface-number by MAC %q: %w", mac, err)
+	}
+
+	// This is guaranteed to avoid conflicting with the main routing table ID
+	// (253-255) because the maximum number of ENI devices on a node is 15 (see
+	// pkg/aws/eni/limits.go). Because the interface number is monotonically
+	// increasing and the lowest available number is reused when devices are
+	// added / removed. This means that the max possible table ID is 25.
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
+}
+
+// retrieveTableIDFromInterfaceNumber returns the table ID based on the
+// interface number. The table ID is the ifindex of the device corresponding to
+// the ENI with the given interface number. This is used for downgrading /
+// using the old ENI datapath.
+func (m *migrator) retrieveTableIDFromInterfaceNumber(ifaceNum int) (int, error) {
+	mac, err := m.getter.GetMACByInterfaceNumber(ifaceNum)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface-number by MAC %q: %w", mac, err)
+	}
+
+	links, err := m.rpdb.LinkList()
+	if err != nil {
+		return -1, fmt.Errorf("failed to list links: %w", err)
+	}
+
+	var (
+		link  netlink.Link
+		found bool
+	)
+	for _, l := range links {
+		if l.Attrs().HardwareAddr.String() == mac {
+			link = l
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return -1, fmt.Errorf("could not find link with MAC %q by interface-number %q", mac, ifaceNum)
+	}
+
+	return link.Attrs().Index, nil
+}
+
+// copyRoutes upserts `routes` under the `from` table ID to `to` table ID. It
+// returns a RevertStack and an error. The RevertStack contains functions that
+// would revert all the successful operations that occurred in this function.
+// The caller of this function should be aware to ensure the stack is reverted
+// upon an error.
```

Comment:
Let's not make this ambiguous ;-)

```suggestion
// The caller of this function MUST revert the stack when this function returns an error.
```

Same for `createNewRule()`.